pkg/tasks/probes.go

```diff
@@ -134,6 +147,139 @@ func safeguard(s *state.State) error {
+			// If there are both, master toleration would be just ignored, so we don't need
+			// to fail.
+			if foundMaster && !foundControlPlane {
+				invalidTolerations = append(invalidTolerations, pod.Name)
+			}
+		}
+
+		var shouldFail bool
+		if len(invalidNodeSelector) > 0 {
+			shouldFail = true
+			s.Logger.Errorf("Found %d pod(s) that are using NodeSelector %q which is removed in Kubernetes 1.24: %s", len(invalidNodeSelector), nodeRoleMaster, invalidNodeSelector)
+		}
+		if len(invalidTolerations) > 0 {
+			shouldFail = true
+			s.Logger.Errorf("Found %d pod(s) that have toleration removed in Kubernetes 1.24 %q, but not %q toleration: %s", len(invalidTolerations), nodeRoleMaster, nodeRoleControlPlane, invalidTolerations)
+		}
+		if shouldFail {
+			s.Logger.Warn("For pods managed by KubeOne, run 'kubeone apply' with your current/actual Kubernetes version before upgrading to Kubernetes 1.24.")
+
+			return fail.RuntimeError{
+				Err: errors.New("invalid tolerations or nodeSelectors"),
+				Op:  "some pods are using tolerations or nodeSelectors removed in Kubernetes 1.24",
+			}
+		}
+	}
+
+	return nil
+}
+
+// safeguardFlatcarMachineDeployments ensures that there are no MachineDeployments
+// running Flatcar Linux and using cloud-init provisioning utility. That's because
+// cloud-init is currently not supported by OSM for Flatcar.
+// This safeguard only warns user about the problem, it doesn't block kubeone apply
+// from reconciling the cluster.
+func safeguardFlatcarMachineDeployments(s *state.State) error {
+	if !s.Cluster.OperatingSystemManager.Deploy || !s.Cluster.MachineController.Deploy {
+		return nil
+	}
+
+	var machinedeployments clusterv1alpha1.MachineDeploymentList
+	if err := s.DynamicClient.List(s.Context, &machinedeployments, dynclient.InNamespace(metav1.NamespaceSystem)); err != nil {
+		return fail.KubeClient(err, "getting machinedeployments")
+	}
+
+	cloudInitMDs := []string{}
+
+	for _, md := range machinedeployments.Items {
+		providerConfig := providerconfigtypes.Config{}
+		if err := jsonutil.StrictUnmarshal(md.Spec.Template.Spec.ProviderSpec.Value.Raw, &providerConfig); err != nil {
+			return fail.RuntimeError{
+				Err: err,
+				Op:  "decoding providerconfig for machinedeployment",
+			}
+		}
+
+		// KubeOne 1.4 has been using cloud-init for Flatcar, but it doesn't work
+		// with OSM, so users have to switch to Ignition.
+		if providerConfig.OperatingSystem == providerconfigtypes.OperatingSystemFlatcar {
+			var osConfig map[string]interface{}
+			if err := json.Unmarshal(providerConfig.OperatingSystemSpec.Raw, &osConfig); err != nil {
+				return fail.RuntimeError{
+					Err: err,
+					Op:  "decoding operatingsystemconfig for machinedeployment",
+				}
+			}
+
+			if v, ok := osConfig[provisioningUtilityKey]; ok && v == provisioningUtilityCloudInit {
+				if aVal := md.Annotations[machineDeploymentOSMAnnotation]; aVal == "" || aVal == ospFlatcar {
+					cloudInitMDs = append(cloudInitMDs, md.Name)
+				}
```

Comment:
We should omit this check. Since for existing machine deployments, the OSP annotation wouldn't exist anyways.

Reply:
My bad, I didn't properly go through the condition before commenting. This looks good, as expected ;)